docs/payload_context.rst

```diff
@@ -0,0 +1,84 @@
+.. _payload_context:
+
+Payload Context
+===============
+
+Sometimes it is beneficial to pass some contextual information from the trigger
+to the trigger listener along the payload. Examples are:
+
+- tracing information that allows to track complex request processing in a
+  multi component system
+- in a multitenant system a tenant information to be able to identify the
+  tenant that peformed an operation that triggered a notification
+
+
+This can be done by using **Payload Context**. This feature includes:
+
+- ability to add an additional information to the payload in the trigger
+- ability to filter by the fields in the context in the listener process
+- ability to use ``context`` fields in the listener callbacks
+
+
+Add ``context`` to payload in the trigger
+-----------------------------------------
+
+Before doing updates that produce notifications set the context that should be
+passed using ``pgpubsub.set_notification_context`` function. 
+
+.. code-block:: python
+
+    from pgpubsub import set_notification_context
+
+    set_notification_context({'some-key': 'some-value'})
+
+The setting is effective till the connection is closed. Alternatively the setting
+``PGPUBSUB_TX_BOUND_NOTIFICATION_CONTEXT`` can be used to clean the context at the end 
+of the current transanction.
+
+
+Filter by ``context`` field in the trigger listener
+---------------------------------------------------
+
+Define a class that implements ``ListenerFilterProvider`` protocol and set
+option ``PGPUBSUB_LISTENER_FILTER`` to its fully qualified class name.
+
+.. code-block:: python
+
+    from pgpubsub import ListenerFilterProvider
+
+    class TenantListenerFilterProvider(ListenerFilterProvider):
+        def get_filter(self) -> Q:
+            return Q(payload__context__tenant='my-tenant')
+
+    # django settings
+    PGPUBSUB_LISTENER_FILTER = 'myapp.whatever.TenantListenerFilterProvider'
```

Comment:
Does this implicitly imply we can only have one filter provider at a time? If so that seems a little restrictive - I would think there could be many cases that we'd want one listener to filter by one key/value, and another by another key/value. I can see how this could be a good global filter if we wanted one, so it's a good start in that regard.

I think a bit more of versatile API would be something like

```python
@pgpubsub.post_insert_listener(AuthorTriggerChannel, context_filter=TenantListenerFilterProvider())
    def create_first_post_for_author(
        old: Author, new: Author, context: Dict[str, Any]
    ):
```

Such an API would allow different filters per listener. It could also be used to imply that listener with non null context_filter is automatically accepting contexts, hence removing the need for PGPUBSUB_PASS_CONTEXT_TO_LISTENERS. Of course PGPUBSUB_PASS_CONTEXT_TO_LISTENERS is good as a global back up as is the other setting.

I guess we can think of this is an addition to what we have here, so not a blocker. Just thought I'd mention to see what you think?

Reply:
That's true that the filter provider is global and the current implementation is restrictive in this regard. I don't see a use-case for non global filter provider. 

The only use case I have at hand and the motivation for this PR is the deployments where one DB is shared by multilple tenant but each tenant has it's own listener process and needs to process only notifications about changes done by that tenant. For this scenario global filter provider works fine. In order to generalize I need at least one real life example how this would be used to provide a useful API.

Re `PGPUBSUB_PASS_CONTEXT_TO_LISTENERS` I think it is possible to get rid of it completely by analyzing the signature of the listener callback.